src/test/pybind/test_rbd.py

```diff
@@ -415,6 +415,18 @@ def progress_cb(current, total):
     assert_raises(OperationCanceled, RBD().remove, ioctx, image_name,
                   on_progress=progress_cb)
 
+def test_remove_with_progress_except():
+    create_image()
+    d = {'received_callback': False}
+    def progress_cb(current, total):
+        d['received_callback'] = True
+        raise Exception()
+        return -ECANCELED
+
+    # exception is ignored
```

Comment:
I did not think you would be so fast to review it :-) I was going to add some explanation after making sure the current version passes tests.

Anyway, here is a test script that I made to work on env built with cython < 3 without my current changes, to see how it currently behaves:
```
#!/usr/bin/env python3

from rados import Rados
from rbd import RBD, Image, PermissionError


IMG_SIZE = 8 << 20 # 8 MiB
IMG_ORDER = 22 # 4 MiB objects

rados = Rados(conffile='')
rados.connect()
pool_name = 'rbd'
ioctx = rados.open_ioctx(pool_name)
RBD().pool_init(ioctx, True)
image_name = 'testXYZAJDYE765'

try:
    RBD().create(ioctx, image_name, IMG_SIZE, IMG_ORDER)
except:
    print('image already exists')

image = Image(ioctx, image_name)

retval = [None]
def cb(_, buf):
    retval[0] = buf

comp = image.aio_read(0, 20, cb)
comp.wait_for_complete_and_cb()
assert retval[0] == b'\0' * 20
assert comp.get_return_value() == 20

def cb(_, buf):
    raise KeyError('foo')

try:
    comp = image.aio_read(0, 20, cb)
    comp.wait_for_complete_and_cb()
except KeyError as e:
    print(f'KeyError: {e}')

image.close()

d = {'received_callback': False}
def progress_cb(current, total):
    d['received_callback'] = True
    raise Exception()
    return 1

RBD().remove(ioctx, image_name, on_progress=progress_cb)

assert d['received_callback']
```
As you may see, the current behavior for the aio callback to propagate the exception and for the progress callback is to ignore exception. So I adjusted my patch (changed `except? -1` to `except? 0`) to make cython3 compiled module behave the same way.

Reply:
> As you may see, the current behavior for the aio callback to propagate the exception and for the progress callback is to ignore exception.

At first this didn't make sense to me, because Cython 2 is supposed to ignore exceptions universally (by default).  But after looking at the code, I see that there is special code around the aio callback, so it makes sense.  Here the exception is caught and stashed:
https://github.com/ceph/ceph/blob/e89bb53732569ae4a8d78945935b6e5cb35576be/src/pybind/rbd/rbd.pyx#L506-L518
And here it's re-raised from a non-`cdef` function:
https://github.com/ceph/ceph/blob/e89bb53732569ae4a8d78945935b6e5cb35576be/src/pybind/rbd/rbd.pyx#L465-L479

> So I adjusted my patch (changed `except? -1` to `except? 0`) to make cython3 compiled module behave the same way.

Are you sure it made a difference?  AFAIU this just changed the exception indicator from `-1` to `0`, I believe the call to `PyErr_Occurred()` would still be inserted where applicable.  Essentially, it's a change from:
```
ret_val = foo();
if (ret_val == -1 && PyErr_Occurred()) goto error_handler;
```
to
```
ret_val = foo();
if (ret_val == 0 && PyErr_Occurred()) goto error_handler;
```

Reply:
> to make cython3 compiled module behave the same way.

If you going for keeping the existing behavior, I think we need to find a compatible-with-Cython-2 way to mark both of these as `noexcept`.

Reply:
> I left the callback exception test just in case you find them useful.

+1, tests are always useful ;)  Minor nit on this line:
```suggestion
    # exception is logged and ignored with a Cython warning:
    #   Exception ignored in: 'rbd.progress_callback'
```